Do not do smooth keyboard scrolling on containers with scroll snaps

danae404 opened the pull request:
#### 4334fd59e1e2fee0b7ee00d69541db9b04fc5ef0
<pre>
Do not do smooth keyboard scrolling on containers with scroll snaps
<a href="https://bugs.webkit.org/show_bug.cgi?id=242523">https://bugs.webkit.org/show_bug.cgi?id=242523</a>
rdar://96681582

Reviewed by Tim Horton.

Tests:
LayoutTests/tiled-drawing/scrolling/scroll-snap/scroll-snap-mandatory-mainframe-
horizontal-with-keyboard-smooth-scroll-enabled.html
LayoutTests/tiled-drawing/scrolling/scroll-snap/scroll-snap-mandatory-mainframe-
vertical-with-keyboard-smooth-scroll-enabled.html

Currently, smooth keyboard scrolling does not work with scroll snapping. This
patch makes it so when the focused scrollable container uses scroll snapping,
smooth keyboard scrolling is not used. Anywhere where it is checked that
smooth scrolling is enabled, it is now additionally checked that the focused
container does not have scroll snaps.

Before this change, when the pageUp or pageDown key was pressed, it was checked
whether smooth scroll was enabled in WKWebViewMac. To make for easier access to
EventHandler::focusedScrollableAreaUsesScrollSnap, checking whether to do smooth
scroll for the pageUp and pageDown keys is now performed in WebCore::EditorCommand.

Additionally, this patch refactors the smooth keyboard scrolling code so that
EventHandler::keyboardScrollRecursively takes ScrollDirection and ScrollGranularity
As parameters instead of a KeyboardEvent. This makes it simpler to call
EventHandler::keyboardScrollRecursively from EditorCommand because or else an
Artificial KeyboardEvent would need to be created.

* LayoutTests/tiled-drawing/scrolling/scroll-snap/scroll-snap-mandatory-mainframe-
horizontal-with-keyboard-smooth-scroll-enabled-expected.txt: Added.
* LayoutTests/tiled-drawing/scrolling/scroll-snap/scroll-snap-mandatory-mainframe-
horizontal-with-keyboard-smooth-scroll-enabled.html: Added.

Tests that when smooth scroll is turned on, scroll snapping works with the upArrow,
downArrow, pageUp, and pageDown keys.

* LayoutTests/tiled-drawing/scrolling/scroll-snap/scroll-snap-mandatory-mainframe-
vertical-with-keyboard-smooth-scroll-enabled-expected.txt: Added.
* LayoutTests/tiled-drawing/scrolling/scroll-snap/scroll-snap-mandatory-mainframe-
vertical-with-keyboard-smooth-scroll-enabled.html: Added.

Tests that when smooth scroll is turned on, scroll snapping works with the leftArrow
and rightArrow keys.

LayoutTests/TestExpectations
LayoutTests/fast/scrolling/keyboard-scrolling-last-timestamp-expected.txt: Deleted.
LayoutTests/fast/scrolling/keyboard-scrolling-last-timestamp.html: Deleted.
LayoutTests/platform/wk2/TestExpectations

This test was added by bugs.webkit.org/show_bug.cgi?id=229784. This patch
Reverts the changes to pageUp/pageDown scrolling introduced by it,
So the test no longer applies.

* Source/WebCore/editing/EditorCommand.cpp:
(WebCore::EditorCommand::executeScrollPageBackward):
(WebCore::EditorCommand::executeScrollPageForward):

If smooth scroll is enabled and focused container has no scroll snaps, calls
EventHandler::keyboardScrollRecursively, which starts the smooth scroll animation.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::defaultKeyboardEventHandler):

Refactoring of pageUp and pageDown scrolling.

(WebCore::EventHandler::defaultSpaceEventHandler):
(WebCore::EventHandler::defaultArrowEventHandler):

Now call above two functions to obtain arguments to pass to keyboardScrollRecursively.

(WebCore::EventHandler::beginKeyboardScrollGesture):
(WebCore::EventHandler::startKeyboardScrollAnimationOnDocument):
(WebCore::EventHandler::startKeyboardScrollAnimationOnRenderBoxLayer):
(WebCore::EventHandler::startKeyboardScrollAnimationOnRenderBoxAndItsAncestors):
(WebCore::EventHandler::startKeyboardScrollAnimationOnEnclosingScrollableContainer):
(WebCore::EventHandler::keyboardScrollRecursively):

The above functions were refactored to take ScrollDirection and ScrollGranularity as
Parameters rather than a KeyboardEvent.

(WebCore::EventHandler::focusedScrollableAreaUsesScrollSnap):

Returns true if focused container has scroll snaps.

(WebCore::EventHandler::shouldUseSmoothKeyboardScrollingOnFocusedScrollableArea):

Returns true if setting is turned on and focused container does not have scroll snaps.

(WebCore::EventHandler::keyboardScrollRecursively):

Now checks if focused container has scroll snaps.

* Source/WebCore/page/EventHandler.h:

* Source/WebCore/platform/KeyboardScrollingAnimator.cpp:

(WebCore::KeyboardScrollingAnimator::keyboardScrollingKeyForKeyboardEvent): Added.

Helper function for scrollDirectionForKeyboardEvent and scrollGranularityForKeyboardEvent.

(WebCore::KeyboardScrollingAnimator::scrollDirectionForKeyboardEvent): Added.
(WebCore::KeyboardScrollingAnimator::scrollGranularityForKeyboardEvent): Added.

These functions are called by callers of EventHandler::keyboardScrollRecursively.

(WebCore::KeyboardScrollingAnimator::keyboardScrollingKeyFromEvent): Deleted.
(WebCore::KeyboardScrollingAnimator::makeKeyboardScroll): Added.
(WebCore::KeyboardScrollingAnimator::keyboardScrollForKeyboardEvent): Deleted.
(WebCore::KeyboardScrollingAnimator::beginKeyboardScrollGesture):

Refactored to not use KeyboardEvent&apos;s.

Source/WebCore/platform/KeyboardScrollingAnimator.h:

* Source/WebCore/platform/ScrollAnimator.h:
(Webcore::ScrollAnimator::usesScrollSnap):

Gets m_scrollController.usesScrollSnap() for use by
EventHandler::focusedScrollableAreaUsesScrollSnap.

* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebViewMac scrollPageDown:]): Deleted.
(-[WKWebViewMac scrollPageUp:]): Deleted.

Refactoring of pageUp and pageDown scrolling.

* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::performNonEditingBehaviorForSelector):

Now checks if focused container has scroll snaps.

Canonical link: <a href="https://commits.webkit.org/252663@main">https://commits.webkit.org/252663@main</a>
</pre>

hortont424 commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -4283,6 +4281,93 @@ void EventHandler::defaultTextInputEventHandler(TextEvent& event)
         event.setDefaultHandled();
 }
 
+static const std::optional<KeyboardScrollingKey> keyboardScrollingKeyForKeyboardEvent(const KeyboardEvent& event)
```
I think _ideally_ these would stay in KeyboardScrollingAnimator (if you need to use them from here, you can expose them as static methods on KeyboardScrollingAnimator, or as free functions). But maybe there is some reason they moved that I haven't realized yet?

hortont424 replied:
(and that will make this diff much smaller)

danae404 replied:
Fixed

hortont424 commented on Source/WebCore/platform/KeyboardScrollingAnimator.h:
```diff
@@ -33,20 +34,34 @@ namespace WebCore {
     void stopKeyboardScrollAnimation();
     RectEdges<bool> scrollableDirectionsFromPosition(FloatPoint) const;
-    std::optional<KeyboardScroll> keyboardScrollForKeyboardEvent(const PlatformKeyboardEvent&) const;
+    std::optional<KeyboardScroll> getKeyboardScroll(ScrollDirection, ScrollGranularity) const;
```
We use the prefix "get" for things that have out-arguments, which this does not. (maybe makeKeyboardScroll? or maybe it should just be a constructor on KeyboardScroll?)

hortont424 commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -4303,7 +4299,7 @@ void EventHandler::defaultSpaceEventHandler(KeyboardEvent& event)
         return;
 
-    bool defaultHandled = m_frame.settings().eventHandlerDrivenSmoothKeyboardScrollingEnabled() ? keyboardScrollRecursively(event, nullptr) : view->logicalScroll(direction, ScrollGranularity::Page);
+    bool defaultHandled = (m_frame.settings().eventHandlerDrivenSmoothKeyboardScrollingEnabled() && !focusedScrollableAreaUsesScrollSnap()) ? keyboardScrollRecursively(scrollDirectionForKeyboardEvent(event), scrollGranularityForKeyboardEvent(event), nullptr) : view->logicalScroll(direction, ScrollGranularity::Page);
```
This line is WILDLY long, can we rewrite it maybe like

```
bool defaultHandled = false;
if (m_frame.settings().eventHandlerDrivenSmoothKeyboardScrollingEnabled() && !focusedScrollableAreaUsesScrollSnap())
    defaultHandled = keyboardScrollRecursively(scrollDirectionForKeyboardEvent(event), scrollGranularityForKeyboardEvent(event), nullptr);
else ...

hortont424 commented on Source/WebCore/editing/EditorCommand.cpp:
```diff
@@ -1013,11 +1013,17 @@ static bool executeRemoveFormat(Frame& frame, Event*, EditorCommandSource, const
 
 static bool executeScrollPageForward(Frame& frame, Event*, EditorCommandSource, const String&)
 {
+    if (frame.settings().eventHandlerDrivenSmoothKeyboardScrollingEnabled() && !frame.eventHandler().focusedScrollableAreaUsesScrollSnap())
```
I wonder if eventHandler should have a getter that checks both of these things, since you have to check both together always? (EventHandler::shouldUseSmoothKeyboardScrollingForFocusedScrollableArea or something?)

hortont424 replied:
Wouldn't want someone to forget the scroll snap check somewhere...

hortont424 commented on Source/WebCore/platform/KeyboardScrollingAnimator.cpp:
```diff
@@ -96,6 +85,94 @@ static FloatSize perpendicularAbsoluteUnitVector(ScrollDirection direction)
+    if (!key)
+        return { };
+
+    // FIXME (bug 227459): This logic does not account for writing-mode.
```
I can't quite work out why the bodies of these methods are showing up as changes? Did they just get moved around in the file?

danae404 replied:
They used to be lambda expressions in KeyboardScrollingAnimator::keyboardScrollForKeyboardEvent. Separating them into two separate functions introduced a couple of minute changes. Should I try to do something to reduce the size of the diff for this file?

hortont424 commented on Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
```diff
@@ -434,7 +434,7 @@ static String commandNameForSelectorName(const String& selectorName)
     bool didPerformAction = false;
     
-    if (!frame->settings().eventHandlerDrivenSmoothKeyboardScrollingEnabled()) {
+    if (!frame->settings().eventHandlerDrivenSmoothKeyboardScrollingEnabled() || frame->eventHandler().focusedScrollableAreaUsesScrollSnap()) {
```
I think this can adopt your new combo getter, right?

danae404 replied:
Oops, missed that. Will fix!

danae404 committed:
Do not do smooth keyboard scrolling on containers with scroll snaps
https://bugs.webkit.org/show_bug.cgi?id=242523
rdar://96681582

Reviewed by Tim Horton.

Tests:
LayoutTests/tiled-drawing/scrolling/scroll-snap/scroll-snap-mandatory-mainframe-
horizontal-with-keyboard-smooth-scroll-enabled.html
LayoutTests/tiled-drawing/scrolling/scroll-snap/scroll-snap-mandatory-mainframe-
vertical-with-keyboard-smooth-scroll-enabled.html

Currently, smooth keyboard scrolling does not work with scroll snapping. This
patch makes it so when the focused scrollable container uses scroll snapping,
smooth keyboard scrolling is not used. Anywhere where it is checked that
smooth scrolling is enabled, it is now additionally checked that the focused
container does not have scroll snaps.

Before this change, when the pageUp or pageDown key was pressed, it was checked
whether smooth scroll was enabled in WKWebViewMac. To make for easier access to
EventHandler::focusedScrollableAreaUsesScrollSnap, checking whether to do smooth
scroll for the pageUp and pageDown keys is now performed in WebCore::EditorCommand.

Additionally, this patch refactors the smooth keyboard scrolling code so that
EventHandler::keyboardScrollRecursively takes ScrollDirection and ScrollGranularity
As parameters instead of a KeyboardEvent. This makes it simpler to call
EventHandler::keyboardScrollRecursively from EditorCommand because or else an
Artificial KeyboardEvent would need to be created.

* LayoutTests/tiled-drawing/scrolling/scroll-snap/scroll-snap-mandatory-mainframe-
horizontal-with-keyboard-smooth-scroll-enabled-expected.txt: Added.
* LayoutTests/tiled-drawing/scrolling/scroll-snap/scroll-snap-mandatory-mainframe-
horizontal-with-keyboard-smooth-scroll-enabled.html: Added.

Tests that when smooth scroll is turned on, scroll snapping works with the upArrow,
downArrow, pageUp, and pageDown keys.

* LayoutTests/tiled-drawing/scrolling/scroll-snap/scroll-snap-mandatory-mainframe-
vertical-with-keyboard-smooth-scroll-enabled-expected.txt: Added.
* LayoutTests/tiled-drawing/scrolling/scroll-snap/scroll-snap-mandatory-mainframe-
vertical-with-keyboard-smooth-scroll-enabled.html: Added.

Tests that when smooth scroll is turned on, scroll snapping works with the leftArrow
and rightArrow keys.

LayoutTests/TestExpectations
LayoutTests/fast/scrolling/keyboard-scrolling-last-timestamp-expected.txt: Deleted.
LayoutTests/fast/scrolling/keyboard-scrolling-last-timestamp.html: Deleted.
LayoutTests/platform/wk2/TestExpectations

This test was added by bugs.webkit.org/show_bug.cgi?id=229784. This patch
Reverts the changes to pageUp/pageDown scrolling introduced by it,
So the test no longer applies.

* Source/WebCore/editing/EditorCommand.cpp:
(WebCore::EditorCommand::executeScrollPageBackward):
(WebCore::EditorCommand::executeScrollPageForward):

If smooth scroll is enabled and focused container has no scroll snaps, calls
EventHandler::keyboardScrollRecursively, which starts the smooth scroll animation.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::defaultKeyboardEventHandler):

Refactoring of pageUp and pageDown scrolling.

(WebCore::EventHandler::defaultSpaceEventHandler):
(WebCore::EventHandler::defaultArrowEventHandler):

Now call above two functions to obtain arguments to pass to keyboardScrollRecursively.

(WebCore::EventHandler::beginKeyboardScrollGesture):
(WebCore::EventHandler::startKeyboardScrollAnimationOnDocument):
(WebCore::EventHandler::startKeyboardScrollAnimationOnRenderBoxLayer):
(WebCore::EventHandler::startKeyboardScrollAnimationOnRenderBoxAndItsAncestors):
(WebCore::EventHandler::startKeyboardScrollAnimationOnEnclosingScrollableContainer):
(WebCore::EventHandler::keyboardScrollRecursively):

The above functions were refactored to take ScrollDirection and ScrollGranularity as
Parameters rather than a KeyboardEvent.

(WebCore::EventHandler::focusedScrollableAreaUsesScrollSnap):

Returns true if focused container has scroll snaps.

(WebCore::EventHandler::shouldUseSmoothKeyboardScrollingOnFocusedScrollableArea):

Returns true if setting is turned on and focused container does not have scroll snaps.

(WebCore::EventHandler::keyboardScrollRecursively):

Now checks if focused container has scroll snaps.

* Source/WebCore/page/EventHandler.h:

* Source/WebCore/platform/KeyboardScrollingAnimator.cpp:

(WebCore::KeyboardScrollingAnimator::keyboardScrollingKeyForKeyboardEvent): Added.

Helper function for scrollDirectionForKeyboardEvent and scrollGranularityForKeyboardEvent.

(WebCore::KeyboardScrollingAnimator::scrollDirectionForKeyboardEvent): Added.
(WebCore::KeyboardScrollingAnimator::scrollGranularityForKeyboardEvent): Added.

These functions are called by callers of EventHandler::keyboardScrollRecursively.

(WebCore::KeyboardScrollingAnimator::keyboardScrollingKeyFromEvent): Deleted.
(WebCore::KeyboardScrollingAnimator::makeKeyboardScroll): Added.
(WebCore::KeyboardScrollingAnimator::keyboardScrollForKeyboardEvent): Deleted.
(WebCore::KeyboardScrollingAnimator::beginKeyboardScrollGesture):

Refactored to not use KeyboardEvent's.

Source/WebCore/platform/KeyboardScrollingAnimator.h:

* Source/WebCore/platform/ScrollAnimator.h:
(Webcore::ScrollAnimator::usesScrollSnap):

Gets m_scrollController.usesScrollSnap() for use by
EventHandler::focusedScrollableAreaUsesScrollSnap.

* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebViewMac scrollPageDown:]): Deleted.
(-[WKWebViewMac scrollPageUp:]): Deleted.

Refactoring of pageUp and pageDown scrolling.

* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::performNonEditingBehaviorForSelector):

Now checks if focused container has scroll snaps.

Canonical link: https://commits.webkit.org/252663@main
